File: src/lib/backup/ports.rs
use super::model::BackupError;
use std::path::Path;

pub trait BackupService {
    /// Creates a backup of the specified file
    ///
    /// # Arguments
    /// * `file_path` - Path to the file to backup
    ///
    /// # Returns
    /// * `Ok(None)` - File doesn't exist (new file, no backup needed)
    /// * `Ok(Some(path))` - Backup created successfully, returns path to backup
    /// * `Err(e)` - Failed to create backup
    fn create_backup(&self, file_path: &Path) -> Result<Option<String>, BackupError>;
}

File: pkg/backup/ports.go
package backup

// Service is the interface for backup operations. It is the port of the Rust
// trait `BackupService`.
type Service interface {
	// CreateBackup creates a backup of the specified file.
	//
	// Arguments:
	//   - filePath: path to the file to backup
	//
	// Returns:
	//   - (nil, nil): file doesn't exist (new file, no backup needed) (was Ok(None))
	//   - (path, nil): backup created successfully, returns path to backup
	//   - (nil, err): failed to create backup
	CreateBackup(filePath string) (*string, error)
}
